lib/events/auditlog.go

```diff
@@ -0,0 +1,576 @@
+		// optimization: to avoid parsing JSON unnecessarily, lets see if we
+		// can filter out lines that don't even have the requested event type on the line
+		for i := range eventFilter {
+			if strings.Contains(scanner.Text(), eventFilter[i]) {
+				accepted = true
+				break
+			}
+		}
+		if doFilter && !accepted {
+			continue
+		}
+		// parse JSON on the line and compare event type field to what's
+		// in the query:
+		var ef EventFields
+		if err = json.Unmarshal(scanner.Bytes(), &ef); err != nil {
+			logrus.Warnf("invalid JSON in %s line %d", fn, lineNo)
+		}
+		for i := range eventFilter {
+			if ef.GetString(EventType) == eventFilter[i] {
+				accepted = true
+				break
+			}
+		}
+		if accepted || !doFilter {
+			retval = append(retval, ef)
+		}
+	}
+	return retval, nil
+}
+
+// rotateLog() checks if the current log file is older than a given duration,
+// and if it is, closes it and opens a new one
+func (l *AuditLog) rotateLog() (err error) {
+	// determine the timestamp for the current log file
+	fileTime := l.TimeSource().In(time.UTC).Round(l.RotationPeriod)
+
+	openLogFile := func() error {
+		l.Lock()
+		defer l.Unlock()
+		logfname := filepath.Join(l.dataDir,
+			fileTime.Format("2006-01-02.15:04:05")+LogfileExt)
+		l.file, err = os.OpenFile(logfname, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0640)
+		if err != nil {
+			logrus.Error(err)
+		}
+		l.fileTime = fileTime
+		return trace.Wrap(err)
+	}
+
+	// need to create a log file?
+	if l.file == nil {
+		return openLogFile()
+	}
+
+	// time to advance the logfile?
+	if l.fileTime.Before(fileTime) {
+		l.file.Close()
+		return openLogFile()
+	}
+	return nil
+}
+
+// Closes the audit log, which inluces closing all file handles and releasing
+// all session loggers
+func (l *AuditLog) Close() error {
+	l.Lock()
+	defer l.Unlock()
+	if l.file != nil {
+		l.file.Close()
+		l.file = nil
+	}
+	for sid, logger := range l.loggers {
+		logger.Close()
+		delete(l.loggers, sid)
+	}
+	return nil
+}
+
+// sessionStreamFn helper determins the name of the stream file for a given
+// session by its ID
+func (l *AuditLog) sessionStreamFn(sid session.ID) string {
+	return filepath.Join(
+		l.dataDir,
+		SessionLogsDir,
+		fmt.Sprintf("%s%s", sid, SessionStreamPrefix))
+}
+
+// sessionLogFn helper determins the name of the stream file for a given
+// session by its ID
+func (l *AuditLog) sessionLogFn(sid session.ID) string {
+	return filepath.Join(
+		l.dataDir,
+		SessionLogsDir,
+		fmt.Sprintf("%s%s", sid, SessionLogPrefix))
+}
+
+// LoggerFor creates a logger for a specified session. Session loggers allow
+// to group all events into special "session log files" for easier audit
+func (l *AuditLog) LoggerFor(sid session.ID) (sl *SessionLogger) {
+	l.Lock()
+	defer l.Unlock()
+
+	sl, ok := l.loggers[sid]
+	if ok {
+		return sl
+	}
+	// make sure session logs dir is present
+	sdir := filepath.Join(l.dataDir, SessionLogsDir)
+	if err := os.MkdirAll(sdir, 0770); err != nil {
+		logrus.Error(err)
```

Comment:
in case if logger does not exist, return `trace.NotFound` and let caller handle this case gracefully